sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala

```diff
@@ -0,0 +1,254 @@
+      firstRowInNextGroup = inputRow.copy()
+      sortedInputHasNewGroup = true
+    } else {
+      // This inputIter is empty.
+      sortedInputHasNewGroup = false
+    }
+  }
+
+  initialize()
+
+  /** Processes rows in the current group. It will stop when it find a new group. */
+  protected def processCurrentSortedGroup(): Unit = {
+    currentGroupingKey = nextGroupingKey
+    currentSession = nextGroupingSession
+
+    // Now, we will start to find all rows belonging to this group.
+    // We create a variable to track if we see the next group.
+    var findNextPartition = false
+    // firstRowInNextGroup is the first row of this group. We first process it.
+    processRow(sortBasedAggregationBuffer, firstRowInNextGroup)
+
+    // The search will stop when we see the next group or there is no
+    // input row left in the iter.
+    while (!findNextPartition && inputIterator.hasNext) {
+      // Get the grouping key.
+      val currentRow = inputIterator.next()
+      val groupingKey = groupingWithoutSessionProjection(currentRow)
+
+      val session = sessionProjection(currentRow)
+      val sessionStruct = session.getStruct(0, 2)
+      val sessionStart = getSessionStart(sessionStruct)
+      val sessionEnd = getSessionEnd(sessionStruct)
+
+      // Check if the current row belongs the current input row.
+      if (currentGroupingKey == groupingKey) {
+        if (sessionStart < getSessionStart(currentSession)) {
+          errorOnIterator = true
+          throw new IllegalStateException("Input iterator is not sorted based on session!")
+        } else if (sessionStart <= getSessionEnd(currentSession)) {
+          // expanding session length if needed
+          expandEndOfCurrentSession(sessionEnd)
+          processRow(sortBasedAggregationBuffer, currentRow)
+        } else {
+          // We find a new group.
+          findNextPartition = true
+          startNewSession(currentRow, groupingKey, sessionStruct)
+        }
+      } else {
+        // We find a new group.
+        findNextPartition = true
+        startNewSession(currentRow, groupingKey, sessionStruct)
+      }
+    }
+
+    // We have not seen a new group. It means that there is no new row in the input
+    // iter. The current group is the last group of the iter.
+    if (!findNextPartition) {
+      sortedInputHasNewGroup = false
+    }
+  }
+
+  private def startNewSession(currentRow: InternalRow, groupingKey: UnsafeRow,
+                              sessionStruct: UnsafeRow): Unit = {
+    nextGroupingKey = groupingKey.copy()
+    nextGroupingSession = sessionStruct.copy()
+    firstRowInNextGroup = currentRow.copy()
+  }
+
+  private def getSessionStart(sessionStruct: UnsafeRow): Long = {
+    sessionStruct.getLong(0)
+  }
+
+  private def getSessionEnd(sessionStruct: UnsafeRow): Long = {
+    sessionStruct.getLong(1)
+  }
+
+  def updateSessionEnd(sessionStruct: UnsafeRow, sessionEnd: Long): Unit = {
```

Comment:
only be used once in `expandEndOfCurrentSession`, maybe not necessary to be a function?

Reply:
will inline.